### PR TITLE
[Example] Zero-case test passing

### PR DESCRIPTION
**Pull request by jnm2**
Serves as an example for discussion here:
https://github.com/nunit/nunit/pull/1940#issuecomment-270383243

### CODE REVIEW

**Comment by jnm2** [comment]
Question [moved to gitter](https://gitter.im/nunit/nunit?at=586e9ce07a3f79ef5dde42af).

File: src/NUnitFramework/framework/Internal/Commands/OneTimeSetUpCommand.cs [hunk 1]
```diff
@@ -75,15 +77,18 @@ public override TestResult Execute(TestExecutionContext context)
                     Test.Fixture = _suite.Fixture;
                 }
 
-                for (int i = _setUpTearDown.Count; i > 0; )
-                    _setUpTearDown[--i].RunSetUp(context);
+                if (!(suiteResult.Test is TestFixture && !suiteResult.Test.HasChildren))
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Commands/OneTimeSetUpCommand.cs]: @CharliePoole Maybe this check could be encapsulated into an `ITest` extension method or some other place?

Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Commands/OneTimeSetUpCommand.cs]: Actually `if (suiteResult.Test.HasChildren)` passes all tests too. The rule (that user setup and teardown are not run for 0 child tests) only makes sense in the context of fixtures, right?

File: src/NUnitFramework/framework/Internal/Execution/CompositeWorkItem.cs [hunk 2]
```diff
@@ -128,13 +128,16 @@ protected override void PerformWork()
...
-                                // were skipped, failed or set to inconclusive in one time setup
-                                // unless we are aborting.
-                                if (Context.ExecutionStatus != TestExecutionStatus.AbortRequested)
-                                    PerformOneTimeTearDown();
                             }
+
+                            // Directly execute the OneTimeFixtureTearDown for tests that
+                            // were skipped, failed or set to inconclusive in one time setup
+                            // unless we are aborting.
+                            // This is important to do even when there are no child tests,
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Execution/CompositeWorkItem.cs]: Alright! Sounds like we may have a solution then?
How would you word it? A "for example, theories" part could maybe still be useful.

File: src/NUnitFramework/framework/Internal/Execution/CompositeWorkItem.cs [hunk 3]
```diff
@@ -106,10 +106,14 @@ protected override void PerformWork()
 
                             CreateChildWorkItems();
 
+                            // This is important to do even when there are no child tests.
+                            // For example, the teardown command is how theories get an
+                            // opportunity to fail when there are no test cases. To execute
+                            // teardown commands, we must execute setup commands.
+                            PerformOneTimeSetUp();
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Execution/CompositeWorkItem.cs]: @CharliePoole how is this?

File: src/NUnitFramework/framework/Internal/Commands/OneTimeSetUpCommand.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Commands/OneTimeSetUpCommand.cs]: That does the job, but it does it by having one component (OneTimeSetUpCommand) be aware of the necessary behavior for different types of Suites. See my comments about how this all came about in #1962. I think we want a solution that makes individual suite types responsible for how their result is calculated based on the results of contained tests.

**Comment by CharliePoole** [comment]
I've gone ahead and finished a limited fix in PR #1940 and created a new issue #1962 to deal with the additional structural problem that we discovered by working on this. Some of your code is in the fix and your ideas about alternative ways of doing this are the reason I'm delaying it.

As you work with us, you will notice that the week or two right before a release is not a great time to try to solve fundamental problems. :wink:

File: src/NUnitFramework/framework/Internal/Commands/OneTimeSetUpCommand.cs (hunk 1 above)
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/Commands/OneTimeSetUpCommand.cs]: Excellent. I assume #1962 covers the concerns I had in https://github.com/nunit/nunit/pull/1940#issuecomment-270411026?

**Comment by jnm2** [comment]
> As you work with us, you will notice that the week or two right before a release is not a great time to try to solve fundamental problems

Haha! I did not expect this to get pushed through for this release but I am very happy about the #1933  getting fixed. Thank you!

> your ideas about alternative ways of doing this are the reason I'm delaying it.

Hopefully we can discuss that over at #1962 after the release? I'm not married to any particular solution, but I think there is a good reason to make the test builders responsible to deciding whether a test should become a suite, and not couple that decision to the number of cases collected from all the builders. I'll have more to say if that's okay with you.

**Comment by CharliePoole** [comment]
Yes, let's discuss. I have to admit I never understood your comment abut making test builders responsible... Currently they are responsible. Perhaps you can clarify on #1962.

**Pull request closed without merging**